bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -276,21 +284,23 @@ private HeaderCacheElement getCachedHeader() throws IOException {
    */
   @VisibleForTesting
   boolean doRefresh() {
-    boolean requiresRefresh = false;
     synchronized (isRefreshing) {
       if (!isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good) {
         isRefreshing.set(true);
-        requiresRefresh = true;
+      } else {
+        return false;
       }
     }
-    if (!requiresRefresh) {
-      return false;
-    }
-    HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
-    synchronized (isRefreshing) {
-      headerCache.set(cacheElement);
-      isRefreshing.set(false);
-      isRefreshing.notifyAll();
+    try {
+      HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
+      synchronized (isRefreshing) {
```

Comment:
Does this actually need to be inside the synchronized block? it looks like headerCache.set is atomic anyway...?